plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/service/RulePersistenceService.java

```diff
@@ -0,0 +1,184 @@
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.core.xcontent.DeprecationHandler;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.plugin.wlm.rule.QueryGroupFeatureType;
+import org.opensearch.plugin.wlm.rule.action.GetRuleResponse;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.sort.SortOrder;
+import org.opensearch.transport.client.Client;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.opensearch.autotagging.Rule._ID_STRING;
+
+/**
+ * This class encapsulates the logic to manage the lifecycle of rules at index level
+ * @opensearch.experimental
+ */
+public class RulePersistenceService {
+    /**
+     * The system index name used for storing rules
+     */
+    public static final String RULES_INDEX = ".rules";
+    private final Client client;
+    private final ClusterService clusterService;
+    private static final Logger logger = LogManager.getLogger(RulePersistenceService.class);
+    /**
+     * The maximum number of results allowed per GET request
+     */
+    public static final int MAX_RETURN_SIZE_ALLOWED_PER_GET_REQUEST = 50;
+
+    /**
+     * Constructor for RulePersistenceService
+     * @param clusterService {@link ClusterService} - The cluster service to be used by RulePersistenceService
+     * @param client {@link Settings} - The client to be used by RulePersistenceService
+     */
+    @Inject
+    public RulePersistenceService(final ClusterService clusterService, final Client client) {
+        this.clusterService = clusterService;
+        this.client = client;
+    }
+
+    /**
+     * Entry point for the get rule api logic in persistence service. If id is provided, we only get a single rule.
+     * Otherwise, we get all rules that satisfy the attributeFilters.
+     * @param id - The id of the rule to get.
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param searchAfter - The sort values from the last document of the previous page, used for pagination
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    public void getRule(
+        String id,
+        Map<Attribute, Set<String>> attributeFilters,
+        String searchAfter,
+        ActionListener<GetRuleResponse> listener
+    ) {
+        try (ThreadContext.StoredContext context = getContext()) {
```

Comment:
we should add a comment here why do we need to stash the context for future references